### ISSUE TITLE
Generating error if callable object is formatted as a string?

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#5213]
Pretty frequently I have errors where I forget to call a method and accidentally pass a method object to `%` or `str.format`. Mypy doesn't complain since it's possible to convert a callable object to a string, though usually that's not what I want.

Example:

```py
class Item:
    def name(self) -> str: ...

def f(x: Item) -> None:
    print('name: {}'.format(x.name))  # No error, though should be x.name() 
    print('name: %s' % x.name)  # Similar to above
    foo(str(x.name))  # Ditto
```

This occasionally happens during refactoring as well, when I replace an instance variable with a method.

Here are a few things we could do about this:
1. Special case these errors and generate a warning even though the code is not necessarily wrong. Maybe require an explicit `repr(x.some_method)` to silence the error (`# type: ignore` or a cast to `object` would also work).
2. Introduce a new strictness flag that causes these to be flagged. Not sure if this should be enabled by default.
3. Introduce a new strictness flag that catches these and other cases where there is no runtime type error, but that we suspect might be errors. Passing a string argument when an iterable object is expected comes to mind as a kind of similar issue, and there are probably others.
4. Do nothing, since there is no runtime type error and the code could well be valid.

Option 2 seems too specialized -- this doesn't seem important enough to add a strictness flag. Personally I prefer the more opinionated option 1. We could add a hint about how to silence the error in case the code happens to be correct.

**Comment by pslacerda** [comment]
I would prefeer option 4 because because seems more suited to a custom plugin for pylint.

**Comment by JukkaL** [comment]
Pylint can't generate this error reliably since it doesn't do type inference.

If mypy had a public plugin API, we'd have another option:

5. Implement this as a mypy plugin that users can optionally install. The plugin could have other optional "ad-hoc" checks as well.

Finally, if we'd have finer-grained control of which errors to generate (#3806), it would be easier to justify enabling additional checks by default, since there would be a simple way to disable them.

**Comment by ilevkivskyi** [comment]
TBH, I have never had this problem. However, I several times had a similar, but harder to diagnose problem: `if sym.node.fullname == 'something.Special': ...` An experienced mypy developer will immediately see what is wrong, but I remember I had troubles with this mistake several times. So I would rather add a warning for this case.

**Comment by JukkaL** [comment]
>  So I would rather add a warning for this case.

Funnily enough, I've never had this problem. There is an existing issue for this: https://github.com/python/mypy/issues/1749

**Comment by ilevkivskyi** [comment]
Now that we have error codes, this can be a simple follow-up for https://github.com/python/mypy/pull/7418

**Comment by JukkaL** [comment]
We still don't have a general way of enabling/disabling specific error codes. When that is supported, this could be just an optional check that can be enabled. `--strict` could also enable this.
